### PR TITLE
ArC - improve the performance of intercepted subclasses

### PR DESCRIPTION
**Pull request by mkouba**
- replace Map.get() with getfield inside the overriding method

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java [hunk 1]
```diff
@@ -309,21 +297,24 @@ public ResultHandle apply(List<InterceptorInfo> interceptors) {
...
             DecorationInfo decoration = bean.getDecoratedMethods().get(method);
             MethodDescriptor forwardDescriptor = forwardingMethods.get(methodDescriptor);
+            List<Type> parameters = method.parameters();
 
-            if (interception != null) {
-                String methodId = "m" + methodIdx++;
-                ResultHandle methodIdHandle = constructor.load(methodId);
+            // Each method has a corresponding InterceptedMethodMetadata field
+            FieldCreator metadataField = subclass
+                    .getFieldCreator("arc$" + methodIdx++, InterceptedMethodMetadata.class.getName())
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:304]: So this field exists even for methods that are not intercepted, just decorated?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java [hunk 2]
```diff
@@ -343,29 +334,28 @@ public ResultHandle apply(List<InterceptorInfo> interceptors) {
...
                 MethodCreator decoratedMethod = subclass.getMethodCreator(methodDescriptor);
 
-                ResultHandle[] params = new ResultHandle[method.parameters().size()];
-                for (int i = 0; i < method.parameters().size(); ++i) {
+                ResultHandle[] params = new ResultHandle[parameters.size()];
+                for (int i = 0; i < parameters.size(); ++i) {
                     params[i] = decoratedMethod.getMethodParam(i);
                 }
 
                 // if(!this.bean == null) return super.foo()
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:356]: Comment no longer accurate.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java [hunk 3]
```diff
@@ -656,27 +651,32 @@ private MethodDescriptor createForwardingMethod(ClassCreator subclass, String pr
...
+            paramsHandle = interceptedMethod.loadNull();
+        } else {
+            paramsHandle = interceptedMethod.newArray(Object.class,
+                    interceptedMethod.load(parameters.size()));
+            for (int i = 0; i < parameters.size(); i++) {
+                interceptedMethod.writeArrayValue(paramsHandle, i, interceptedMethod.getMethodParam(i));
+            }
         }
 
         // if(!this.bean == null) return super.foo()
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:675]: Comment no longer accurate.

**Review by Ladicek — APPROVED** [review]
Looks good, added just some minor comments.

I wonder if it's possible for a bean to have so many interceptors/decorators that the number of fields would be so big that a `Map` would be better. Probably not realistic, though.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java (hunk 1 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:304]: What do you mean? The field isn't actually written if you get `interception == null`.

Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:304]: It's not written, but it exists, and I wonder if it has to.

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:304]: Hmm, yea, technically you could move this declaration inside the `if` condition if that's what you mean.

Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:304]: That's what I meant -- just not sure if there's some reason I'm not aware of :-)

**Review by manovotn — APPROVED** [review]
Played with it locally to see generated classes and LGTM.

I just wonder how much you actually gain in terms of performance with this kind of change.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/AbstractInvocationContext.java [hunk 4]
```diff
@@ -15,6 +15,8 @@
 
 abstract class AbstractInvocationContext implements ArcInvocationContext, Supplier<Map<String, Object>> {
 
+    private static Object[] EMPTY_PARAMS = new Object[0];
```
Comment (gastaldi) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/AbstractInvocationContext.java:18]:
```suggestion
    private static final Object[] EMPTY_PARAMS = new Object[0];
```

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java:304]: That's a good point. Of course, we don't need the field for decorated methods...

**Comment by mkouba** [comment]
> I just wonder how much you actually gain in terms of performance with this kind of change.

The microbenchmarks showed ~ 15% improvement in thrpt for [intercepted invocations](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/InterceptorBenchmark.java) and 25 - 50% improvement for [subclass instantiation + invocation](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/SubclassInstantiationBenchmark.java).

**Comment by mkouba** [comment]
> I wonder if it's possible for a bean to have so many interceptors/decorators that the number of fields would be so big that a `Map` would be better. Probably not realistic, though.

@Ladicek 500 methods should not be a problem and I think that in 99.9% cases a bean will have < 500 methods. Also the size of the constructor body or the number of entries in the constant pool would probably exceeded first...

**Comment by Ladicek** [comment]
So with 500 intercepted methods, you'd have 500 fields of a reference type. Assuming 4 bytes per reference field (which is normal on 64-bit JVMs today thanks to compressed OOPs, on normally sized heaps), you have 2000 bytes just for this. That is a fairly big object if you ask me, but looking at some JVM sizing heuristics (such as humongous allocations in G1, or TLAB sizing), it seems it's nothing unusal for JVMs. So, yea, probably nothing to worry about :-)

**Comment by mkouba** [comment]
> So, yea, probably nothing to worry about :-)

Yup, 500 intercepted methods is a corner case and I've never seen a "real" bean of such parameters.. And TBH if there is a bean like this then the 2KB we add is probably the least problematic part of the app ;-).

**Pull request merged**